File: 백준/Gold/16234. 인구 이동/인구 이동.cc
#include<bits/stdc++.h>
using namespace std;
int n, l, r, a[104][104], visited[104][104], sum, cnt;
vector<pair<int, int>> v;
const int dy[] = {0, 1, 0, -1};
const int dx[] = {1, 0, -1, 0};
void dfs(int y, int x){
    visited[y][x] = 1;
    sum += a[y][x];
    v.push_back({y,x});
    for(int i = 0; i < 4; i++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || nx < 0 || ny >= n || nx >= n || visited[ny][nx]) continue;
        if(l <= abs(a[ny][nx] - a[y][x]) && abs(a[ny][nx] - a[y][x]) <= r){
            dfs(ny, nx);
        }
    }
}
int main(){
    cin >> n >> l >> r;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cin >> a[i][j];
        }
    }
    while(true){
        bool flag = 0;
        memset(visited, 0, sizeof(visited));
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                v.clear();
                sum = 0;
                if(!visited[i][j]) dfs(i, j);
                if(v.size() == 1) continue;
                for(auto b : v) {
                    a[b.first][b.second] = sum / v.size();
                    flag = 1;
                }
            }
        }
        if(!flag) break;
        cnt++;
    }
    cout << cnt << '\n';
}

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, L, R;
    static int[][] board, visited;
    static class Point{
        int y; int x;
        Point(int y, int x){
            this.y = y;
            this.x = x;
        }
    }
    static int[] dy ={-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static List<Point> points = new ArrayList<>();
    static int dfs(int y, int x, int[][] visited){
        int ret = 0;
        ret += board[y][x];
        points.add(new Point(y, x));
        for(int d = 0; d < 4; d++){
            int ny = y + dy[d];
            int nx = x + dx[d];
            if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
            int cha = Math.abs(board[ny][nx] - board[y][x]);
            if(visited[ny][nx] != 0 || cha < L || cha > R) continue;
            visited[ny][nx] = 1;
            ret += dfs(ny, nx, visited);
        }
        return ret;
    }
    static boolean go(){
        visited = new int[n][n];
        boolean flag = false;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(visited[i][j] != 0)continue;
                visited[i][j] = 1;
                int sum = dfs(i, j, visited);
                if(points.size() == 1) {
                    visited[i][j] = 0;
                    points.clear();
                    continue;
                }
                flag = true;
                for(Point p : points) board[p.y][p.x] = sum / points.size();
                points.clear();
            }
        }
        return flag;
    }
    public static void main(String[] args) throws IOException {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var out = new PrintWriter(System.out);

        var st = new StringTokenizer(in.readLine());
        n = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        board = new int[n][n];
        visited = new int[n][n];
        for(int i = 0; i < n; i++){
            st = new StringTokenizer(in.readLine());
            for(int j = 0; j < n; j++){
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int ret = 0;
        while(true){
            if(!go()) break;
            ret++;
        }


        out.println(ret);


        out.flush();
        out.close();
    }
}